src/Symfony/Component/TypeInfo/Type.php

```diff
@@ -29,6 +29,14 @@ abstract class Type implements \Stringable
      */
     public function isSatisfiedBy(callable $specification): bool
     {
+        if ($this instanceof WrappingTypeInterface && $this->wrappedTypeIsSatisfiedBy($specification)) {
+            return true;
+        }
+
+        if ($this instanceof CompositeTypeInterface && $this->composedTypesAreSatisfiedBy($specification)) {
```

Comment:
treating all composite types the same way looks wrong to me. Union and intersection definitely don't work the same in a type system.
To me, subclasses should override this method as appropriate instead.

Reply:
`IntersectionType` and `UnionType`, are implementing `composedTypesAreSatisfiedBy` differently. The union needs one of the specification to be true, whereas the intersection needs all specifications to be true. So it might already be good no?